File: src/main/scala/com/zotoh/frwk/util/MetaUtils.scala
/*??
 * COPYRIGHT (C) 2012 CHERIMOIA LLC. ALL RIGHTS RESERVED.
 *
 * THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE,
 * VERSION 2.0 (THE "LICENSE").
 *
 * THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
 * AND LIMITATIONS UNDER THE LICENSE.
 *
 * You should have received a copy of the Apache License
 * along with this distribution; if not, you may obtain a copy of the
 * License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 ??*/

package com.zotoh.frwk
package util

import scala.language.existentials
import scala.collection.mutable

import java.lang.reflect.{Field,InvocationTargetException,Method,Modifier}
import com.zotoh.frwk.util.CoreUtils._

import org.apache.commons.lang3.{StringUtils=>STU}

/**
 * Utility functions for class related or reflection related operations.
 *
 * @author kenl
 *
 */
object MetaUtils {

  type ListOfCzs= List[Class[_]]


  def isChild(base:Class[_], c:Class[_]): Boolean = {
    if (c==null) false else base.isAssignableFrom(c)
  }

  def isChild(base:Class[_], obj:Any): Boolean = {
    if (obj==null) false else isChild(base, obj.getClass)
  }

  def isBoolean(z:Class[_]) = {
    z.getName match {
      case "boolean" | "Boolean" | "java.lang.Boolean" => true
      case _ => false
    }
  }

  def isChar(z:Class[_]) = {
    z.getName match {
      case "char" | "Char" | "java.lang.Char" => true
      case _ => false
    }
  }

  def isInt(z:Class[_]) = {
    z.getName match {
      case "int" | "Int" | "java.lang.Integer" => true
      case _ => false
    }
  }

  def isLong(z:Class[_]) = {
    z.getName match {
      case "long" | "Long" | "java.lang.Long" => true
      case _ => false
    }
  }

  def isFloat(z:Class[_]) = {
    z.getName match {
      case "float" | "Float" | "java.lang.Float" => true
      case _ => false
    }
  }

  def isDouble(z:Class[_]) = {
    z.getName match {
      case "double" | "Double" | "java.lang.Double" => true
      case _ => false
    }
  }

  def isByte(z:Class[_]) = {
    z.getName match {
      case "byte" | "Byte" | "java.lang.Byte" => true
      case _ => false
    }
  }

  def isShort(z:Class[_]) = {
    z.getName match {
      case "short" | "Short" | "java.lang.Short" => true
      case _ => false
    }
  }

  def isString(z:Class[_]) = {
    z.getName match {
      case "String" | "java.lang.String" => true
      case _ => false
    }
  }

  def isBytes(z:Class[_]) = {
    if (z == classOf[Array[Byte]]) true else false
  }

  /**
   * @param z
   * @return
   */
  def forName(z:String,cl:ClassLoader=null) = {
    if (cl ==null) Class.forName(z) else Class.forName(z,true,cl)
  }

  /**
   * Get the classloader used by this object.
   *
   * @return
   */
  def getCZldr(cl:ClassLoader = null) = {
    if (cl==null) Thread.currentThread().getContextClassLoader else cl
  }

  def setCZldr(cl:ClassLoader) {
    tstObjArg("class-loader",cl)
    Thread.currentThread().setContextClassLoader( cl)
  }

  /**
   *
   * @param clazz
   * @param ld
   * @return
   */
  def loadClass(clazz:String, cl:ClassLoader=null) = {
    if (STU.isEmpty(clazz)) null else getCZldr(cl).loadClass(clazz)
  }

  /**
   * Create an object of this class, calling the default constructor.
   *
   * @param clazz
   * @param ldr optional.
   * @return
   */
  def mkRef(clazz:String, cl:ClassLoader=null): Any = {
    if (STU.isEmpty(clazz)) null else mkRef( loadClass(clazz, cl))
  }

  /**
   * Create an object of this class, calling the default constructor.
   *
   * @param c
   * @return
   */
  def mkRef(c:Class[_]): Any = {
    c.getDeclaredConstructor().newInstance()
  }

  /**
   * @param c
   * @return
   */
  def listParents(c:Class[_]) = {
    // since we always add the original class
    val a= collPars(c, Nil) match {
      case x :: tail if tail.length > 0 => tail
      case lst => lst
    }
    a.toSeq
  }

  /**
   * @param c
   * @return
   */
  def listMethods(c:Class[_]) = {
    collMtds(c, 0, Map[String,Method]()).values.toSeq
  }

  /**
   * @param c
   * @return
   */
  def listFields(c:Class[_]) = {
    collFlds(c, 0, Map[String,Field]() ).values.toSeq
  }

  private def collPars(c:Class[_], bin:ListOfCzs): ListOfCzs = {
    val par = c.getSuperclass
    var rc= if (par != null) {
      collPars(par, bin)
    } else {
      bin
    }
    c :: rc
  }

  private def collFlds(c:Class[_], level:Int,
          bin:Map[String,Field]):Map[String,Field] = {

    val flds= c.getDeclaredFields
    val par = c.getSuperclass
    var m= if (par != null) { collFlds(par, level +1, bin) } else { bin }

    // we only want the inherited fields from parents
    (m /: flds) { (rc, f) =>
      val x= f.getModifiers
      if (level > 0 &&
         (Modifier.isStatic(x) || Modifier.isPrivate(x)) ) rc  else {
         rc + (f.getName -> f)
      }
    }

  }

  private def collMtds(c:Class[_], level:Int, bin:Map[String,Method]): Map[String,Method] = {
    val mtds= c.getDeclaredMethods
    val par = c.getSuperclass
    var mp = if (par != null) { collMtds(par, level +1, bin) } else { bin }

    // we only want the inherited methods from parents
    (mp /: mtds) { (rc,m) =>
      val x= m.getModifiers
      if (level > 0 &&
        ( Modifier.isStatic(x) || Modifier.isPrivate(x))) rc else {
        rc + (m.getName -> m)
      }
    }
  }

}

File: src/main/clojure/com/zotoh/frwk/util/metautils.clj
;;
;; COPYRIGHT (C) 2013 CHERIMOIA LLC. ALL RIGHTS RESERVED.
;;
;; THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
;; MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE
;; VERSION 2.0 (THE "LICENSE").
;;
;; THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL
;; BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
;; MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
;;
;; SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
;; AND LIMITATIONS UNDER THE LICENSE.
;;
;; You should have received a copy of the Apache License
;; along with this distribution; if not you may obtain a copy of the
;; License at
;; http://www.apache.org/licenses/LICENSE-2.0
;;

(ns com.zotoh.frwk.util.metautils 
  (:import (java.lang.reflect Field Method Modifier))
  (:require [ com.zotoh.frwk.util.coreutils :as CU ] )
  (:require [ com.zotoh.frwk.util.strutils :as SU ] )
  )

;;
;; Utility functions for class related or reflection related operations.
;;
;; @author kenl
;;

(defn isChildCZ
  "Returns true if clazz is subclass of this base class."
  [basz cz]
  (if (or (nil? basz) (nil? cz)) false (.isAssignableFrom basz cz)))

(defn isChildObj
  "Returns true if object is subclass of this base class."
  [basz obj]
  (if (or (nil? basz) (nil? obj)) false (isChildCZ basz (.getClass obj))))

(defn bytesClass
  "Return the java class for byte[]."
  []
  (Class/forName "[B"))

(defn charsClass
  "Return the java class for char[]."
  []
  (Class/forName "[C"))

(defn isBoolean
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) ["boolean" "Boolean" "java.lang.Boolean"] ) true false))

(defn isChar
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "char" "Char" "java.lang.Char" ] ) true false))

(defn isInt
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "int" "Int" "java.lang.Integer" ] ) true false))

(defn isLong
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "long" "Long" "java.lang.Long" ] ) true false))

(defn isFloat
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "float" "Float" "java.lang.Float" ]) true false ))

(defn isDouble
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "double" "Double" "java.lang.Double" ]) true false))

(defn isByte
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "byte" "Byte" "java.lang.Byte" ]) true false ))

(defn isShort
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "short" "Short" "java.lang.Short" ]) true false))

(defn isString
  ""
  [cz]
  (if (SU/equalsOneOf (.getName cz) [ "String" "java.lang.String" ]) true false))

(defn isBytes
  ""
  [cz]
  (if (= cz (.getClass (byte-array 0))) true false))

(defn forName
  ""
  ( [z] (forName z nil))
  ( [z cl]
    (if (nil? cl) (java.lang.Class/forName z) (java.lang.Class/forName z true cl))))

(defn getCZldr
  ""
  ([] (getCZldr nil))
  ([cl] (if (nil? cl) (.getContextClassLoader (Thread/currentThread)) cl)))

(defn setCZldr
  ""
  [cl]
  (let []
    (CU/tstObjArg "class-loader" cl)
    (.setContextClassLoader (Thread/currentThread) cl)))

(defn loadCZ
  ""
  ( [clazzName] (loadCZ clazzName nil))
  ( [clazzName cl]
    (if (not (SU/hgl clazzName)) nil (.loadClass (getCZldr cl) clazzName))))

(defn ctorObj
  ""
  [cz]
  (let []
    (CU/tstObjArg "java-class" cz)
    (.newInstance (.getDeclaredConstructor cz (make-array Class 0))  (make-array Object 0)  )))

(defn mkObj
  ""
  ([clazzName] (mkObj clazzName nil))
  ([clazzName cl]
   (if (not (SU/hgl clazzName)) nil (ctorObj (loadCZ clazzName cl)))))

(defn listParents
  ""
  [cz]
  (let [ rc (loop [ bin [] par cz ]
              (if (nil? par)
                bin
                (recur (conj bin par) (.getSuperclass par))))  ]
    ;; since we always add the original class, we need to ignore it on return
    (if (> (.size rc) 1) (rest rc) [] )))


(defn- iterXXX
  ""
  [ cz level getDeclXXX bin ]
  (let [ props (getDeclXXX cz) ]
    (reduce (fn [sum m]
              (let [ x (.getModifiers m) ]
                (if (and (> level 0)
                         (or (Modifier/isStatic x) (Modifier/isPrivate x)) )
                  sum
                  (assoc sum (.getName m) m)))) bin props) ))

(defn- listMtds [ cz level ]
  (let [ par (.getSuperclass cz) ]
    (iterXXX cz level #(.getDeclaredMethods %) (if (nil? par) {} (listMtds par (inc level))))))

(defn- listFlds [ cz level ]
  (let [ par (.getSuperclass cz) ]
    (iterXXX cz level #(.getDeclaredFields %) (if (nil? par) {} (listFlds par (inc level))))))

(defn listMethods
  "List all methods belonging to this class, including inherited ones."
  [cz]
  (vals (if (nil? cz) {} (listMtds cz 0 ))))

(defn listFields
  "List all fields belonging to this class, including inherited ones."
  [cz]
  (vals (if (nil? cz) {} (listFlds cz 0 ))))























(def ^:private metautils-eof nil)

